agent/containermetadata/manager.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+
+	"github.com/aws/amazon-ecs-agent/agent/utils/oswrapper"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+const (
+	metadataEnvironmentVariable = "ECS_CONTAINER_METADATA_FILE"
+	inspectContainerTimeout     = 30 * time.Second
+	metadataFile                = "ecs-container-metadata.json"
+	metadataPerm                = 0644
+)
+
+// MetadataStatus specifies the current update status of the metadata file.
+// The purpose of this status is for users to check if the metadata file has
+// reached the stage they need before they read the rest of the file to avoid
+// race conditions (Since the final stage will need to be after the container
+// starts up
+// In the future the metadata may require multiple stages of update and these
+// statuses should amended/appended accordingly.
+type MetadataStatus string
```

Comment:
I believe this was a section I wrote and the reason I used a string rather than enum was to display the actual name of the status in the JSON Marshaling. I believe using an enum would produce numerical codes in the Marshaled output. which I think is less helpful to users checking for this status.

Reply:
You can still have status names in marshaled JSON output with iotas (Go doesn't actually have enums), but you'd need to implement the [`TextMarshaler`](https://golang.org/pkg/encoding/#TextMarshaler) and [`TextUnmarshaler`](https://golang.org/pkg/encoding/#TextUnmarshaler) interfaces.

I would recommend doing this, as you then get stronger assurances about what will be marshaled/unmarshaled.

Reply:
@adnxn can  you address @samuelkarp's comment as well? Also, can this be moved to `containermetadata/types.go` ?

Reply:
gotcha, fixed. using iotas and moved to `containermetadata/types.go`.